### ISSUE TITLE
Extend screenshot to non-canvas layers

### ISSUE THREAD

**Issue by manthey** [OpenGeoscience/geojs#666]
The map.screenshot() function only gets canvas layers right now.

To get svg layers is straightforward, BUT it will require returning a Promise rather than the screenshot right away, as there is no way to ensure synchronous loading.

It is possible to get HTML layers (and the attribution element), by rendering into an svg foreignobject, and then rendering that into a canvas.

**Comment by manthey** [comment]
I tried the rasterizehtml library, but it didn't handle external image properly (so svg tile layers weren't rendered), and other svg layers appeared with an some offset.

I have a simple function that can render svg layers now, but it is asynchronous.

Is there are objection to switching the screenshot to return a promise which would then return its results?

That is, instead of 
```
  var base64PNG = map.screenshot()
```
you would have
```
  map.screenshot().then(function (result) {
    var base64PNG = result;
  });
```

**Comment by manthey** [comment]
A few other possibilities:

We could trigger an event when the screenshot is ready, so that clients could listen for that instead of dealing with the promise.

We could add an option that would ensure the map was idle and rendered before taking the screenshot.

**Comment by jbeezley** [comment]
I would go with both an event and a screenshot.

**Comment by aashish24** [comment]
> I would go with both an event and a screenshot.

+1. so event and promise both is what I am assuming here.

**Comment by jbeezley** [comment]
Yes, `s/screenshot/promise/`, it's been one of those days.

**Comment by manthey** [comment]
Resolved by PR #667.
